File: leetcode/88. Merge Sorted Array.py
'''
Runtime: 36 ms, faster than 92.86% of Python3 online submissions for Merge Sorted Array.
Memory Usage: 14 MB, less than 40.57% of Python3 online submissions for Merge Sorted Array.

시간복잡도: O(n) -> merged array를 합치는 과정이기 때문에 한 번 스캔만으로 가능
공간복잡도: O(1) -> 주어진 공간에 한해서 사용.
'''

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        
        for i in range(n):
            k = i
            while nums1[k] <= nums2[0] and k < m:
                k+=1
            nums1.pop()
            nums1.insert(k, nums2[0])
            nums2 = nums2[1:]
            m+=1

File: leetcode/Java/88. Merge Sorted Array.java
public class 88. Merge Sorted Array {
    /*
    Runtime: 0 ms, faster than 100.00% of Java online submissions for Merge Sorted Array.
    Memory Usage: 42.7 MB, less than 56.85% of Java online submissions for Merge Sorted Array.

    시간복잡도: O(n) -> 끝자리부터 출발해서 하나씩 비교 -> 이미 nums1, 2 둘다 정렬된 상태기 때문에 한번씩만 비교해도 가능
    공간복잡도: O(1) -> 주어진 nums가지고 끝냄.
     */
    class Solution {
        public void merge(int[] nums1, int m, int[] nums2, int n) {
            int tail1 = m-1, tail2 = n-1, fin = m+n-1;
            while (tail1 >=0 && tail2 >=0) {
                nums1[fin--] = (nums1[tail1] > nums2[tail2]) ? nums1[tail1--] : nums2[tail2--];
            }
            while (tail2 >= 0) {
                nums1[fin--] = nums2[tail2--];
            }
        }
    }
    
}
